In [1]:
import pandas as pd
import env

# Read the data from the attendance table and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

You should end up with something like this:


In [15]:
'''
name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64
'''


'\nname\nBilly    0.5250\nJane     0.6875\nJohn     0.9125\nSally    0.7625\nName: grade, dtype: float64\n'

In [2]:
df = pd.read_csv('attendance.csv')
df

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  4 non-null      object
 1   2018-01-01  4 non-null      object
 2   2018-01-02  4 non-null      object
 3   2018-01-03  4 non-null      object
 4   2018-01-04  4 non-null      object
 5   2018-01-05  4 non-null      object
 6   2018-01-06  4 non-null      object
 7   2018-01-07  4 non-null      object
 8   2018-01-08  4 non-null      object
dtypes: object(9)
memory usage: 416.0+ bytes


In [4]:
df_melt = df.melt(id_vars = df.columns[0])
df_melt

Unnamed: 0    variable value
0       Sally  2018-01-01     P
1        Jane  2018-01-01     A
2       Billy  2018-01-01     A
3        John  2018-01-01     P
4       Sally  2018-01-02     T
5        Jane  2018-01-02     P
6       Billy  2018-01-02     T
7        John  2018-01-02     T
8       Sally  2018-01-03     T
9        Jane  2018-01-03     T
10      Billy  2018-01-03     A
11       John  2018-01-03     H
12      Sally  2018-01-04     H
13       Jane  2018-01-04     T
14      Billy  2018-01-04     A
15       John  2018-01-04     P
16      Sally  2018-01-05     P
17       Jane  2018-01-05     T
18      Billy  2018-01-05     H
19       John  2018-01-05     P
20      Sally  2018-01-06     A
21       Jane  2018-01-06     T
22      Billy  2018-01-06     T
23       John  2018-01-06     T
24      Sally  2018-01-07     T
25       Jane  2018-01-07     A
26      Billy  2018-01-07     P
27       John  2018-01-07     P
28      Sally  2018-01-08     T
29       Jane  2018-01-08     T
30      Billy  2018-01-08     T
31       John  2018-01-08     P

In [5]:
df_melt.rename(columns = {'Unnamed: 0':'Name', 'variable':'date'}, inplace=True)
df_melt

Name        date value
0   Sally  2018-01-01     P
1    Jane  2018-01-01     A
2   Billy  2018-01-01     A
3    John  2018-01-01     P
4   Sally  2018-01-02     T
5    Jane  2018-01-02     P
6   Billy  2018-01-02     T
7    John  2018-01-02     T
8   Sally  2018-01-03     T
9    Jane  2018-01-03     T
10  Billy  2018-01-03     A
11   John  2018-01-03     H
12  Sally  2018-01-04     H
13   Jane  2018-01-04     T
14  Billy  2018-01-04     A
15   John  2018-01-04     P
16  Sally  2018-01-05     P
17   Jane  2018-01-05     T
18  Billy  2018-01-05     H
19   John  2018-01-05     P
20  Sally  2018-01-06     A
21   Jane  2018-01-06     T
22  Billy  2018-01-06     T
23   John  2018-01-06     T
24  Sally  2018-01-07     T
25   Jane  2018-01-07     A
26  Billy  2018-01-07     P
27   John  2018-01-07     P
28  Sally  2018-01-08     T
29   Jane  2018-01-08     T
30  Billy  2018-01-08     T
31   John  2018-01-08     P

In [9]:
df_melt.rename(columns = {'variable':'date'}, inplace=True)
df_melt

Name        date value  count
0   Sally  2018-01-01     P      1
1    Jane  2018-01-01     A      1
2   Billy  2018-01-01     A      1
3    John  2018-01-01     P      1
4   Sally  2018-01-02     T      1
5    Jane  2018-01-02     P      1
6   Billy  2018-01-02     T      1
7    John  2018-01-02     T      1
8   Sally  2018-01-03     T      1
9    Jane  2018-01-03     T      1
10  Billy  2018-01-03     A      1
11   John  2018-01-03     H      1
12  Sally  2018-01-04     H      1
13   Jane  2018-01-04     T      1
14  Billy  2018-01-04     A      1
15   John  2018-01-04     P      1
16  Sally  2018-01-05     P      1
17   Jane  2018-01-05     T      1
18  Billy  2018-01-05     H      1
19   John  2018-01-05     P      1
20  Sally  2018-01-06     A      1
21   Jane  2018-01-06     T      1
22  Billy  2018-01-06     T      1
23   John  2018-01-06     T      1
24  Sally  2018-01-07     T      1
25   Jane  2018-01-07     A      1
26  Billy  2018-01-07     P      1
27   John  2018-01-07     P      1
28  Sally  2018-01-08     T      1
29   Jane  2018-01-08     T      1
30  Billy  2018-01-08     T      1
31   John  2018-01-08     P      1

In [10]:
df_melt = df_melt.replace(to_replace = ['P', 'A', 'T', 'H'], value = [1, 0, .9, .5])
df_melt

Name        date  value
0   Sally  2018-01-01    1.0
1    Jane  2018-01-01    0.0
2   Billy  2018-01-01    0.0
3    John  2018-01-01    1.0
4   Sally  2018-01-02    0.9
5    Jane  2018-01-02    1.0
6   Billy  2018-01-02    0.9
7    John  2018-01-02    0.9
8   Sally  2018-01-03    0.9
9    Jane  2018-01-03    0.9
10  Billy  2018-01-03    0.0
11   John  2018-01-03    0.5
12  Sally  2018-01-04    0.5
13   Jane  2018-01-04    0.9
14  Billy  2018-01-04    0.0
15   John  2018-01-04    1.0
16  Sally  2018-01-05    1.0
17   Jane  2018-01-05    0.9
18  Billy  2018-01-05    0.5
19   John  2018-01-05    1.0
20  Sally  2018-01-06    0.0
21   Jane  2018-01-06    0.9
22  Billy  2018-01-06    0.9
23   John  2018-01-06    0.9
24  Sally  2018-01-07    0.9
25   Jane  2018-01-07    0.0
26  Billy  2018-01-07    1.0
27   John  2018-01-07    1.0
28  Sally  2018-01-08    0.9
29   Jane  2018-01-08    0.9
30  Billy  2018-01-08    0.9
31   John  2018-01-08    1.0

In [13]:
df_melt.groupby('Name').sum()/8

value
Name         
Billy  0.5250
Jane   0.6875
John   0.9125
Sally  0.7625

# 2. Coffee Levels

### a. Read the coffee_levels table.

In [2]:
coffee = pd.read_csv('coffee_levels.csv')
coffee

hour coffee_carafe  coffee_amount
0      8             x       0.816164
1      9             x       0.451018
2     10             x       0.843279
3     11             x       0.335533
4     12             x       0.898291
5     13             x       0.310711
6     14             x       0.507288
7     15             x       0.215043
8     16             x       0.183891
9     17             x       0.391560
10     8             y       0.189297
11     9             y       0.521502
12    10             y       0.023163
13    11             y       0.235529
14    12             y       0.017009
15    13             y       0.997464
16    14             y       0.058361
17    15             y       0.144644
18    16             y       0.544676
19    17             y       0.594126
20     8             z       0.999264
21     9             z       0.915990
22    10             z       0.144928
23    11             z       0.311495
24    12             z       0.771947
25    13             z       0.398520
26    14             z       0.864464
27    15             z       0.436364
28    16             z       0.280621
29    17             z       0.436677

### b. Transform the data so that each carafe is in it's own column.

In [3]:
coffee = coffee.pivot_table(index = 'hour', columns = 'coffee_carafe', values = 'coffee_amount').reset_index()
coffee

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

# 3. Cake Recipes
### a. Read the cake_recipes table. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [3]:
cake = pd.read_csv('cake_recipes.csv')
cake

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

### b. Tidy the data as necessary.

In [4]:
cake['recipe:position'].str.split(':', expand = True)

0       1
0  a  bottom
1  a     top
2  b  bottom
3  b     top
4  c  bottom
5  c     top
6  d  bottom
7  d     top

In [5]:
cake[['recipe_letter', 'position']] = cake['recipe:position'].str.split(':', expand = True)
cake

recipe:position        225        250        275        300 recipe_letter  \
0        a:bottom  61.738655  53.912627  74.414730  98.786784             a   
1           a:top  51.709751  52.009735  68.576858  50.228470             a   
2        b:bottom  57.095320  61.904369  61.196980  99.248541             b   
3           b:top  82.455004  95.224151  98.594881  58.169349             b   
4        c:bottom  96.470207  52.001358  92.893227  65.473084             c   
5           c:top  71.306308  82.795477  92.098049  53.960273             c   
6        d:bottom  52.799753  58.670419  51.747686  56.183110             d   
7           d:top  96.873178  76.101363  59.571620  50.971626             d   

  position  
0   bottom  
1      top  
2   bottom  
3      top  
4   bottom  
5      top  
6   bottom  
7      top

In [18]:
cake_melt = cake.melt(id_vars = ['recipe_letter', 'position'], 
                     var_name = 'temp',
                     value_name = 'score')
cake_melt = cake_melt[8:]
cake_melt = cake_melt.reset_index()
cake_melt

index recipe_letter position temp      score
0       8             a   bottom  225  61.738655
1       9             a      top  225  51.709751
2      10             b   bottom  225   57.09532
3      11             b      top  225  82.455004
4      12             c   bottom  225  96.470207
5      13             c      top  225  71.306308
6      14             d   bottom  225  52.799753
7      15             d      top  225  96.873178
8      16             a   bottom  250  53.912627
9      17             a      top  250  52.009735
10     18             b   bottom  250  61.904369
11     19             b      top  250  95.224151
12     20             c   bottom  250  52.001358
13     21             c      top  250  82.795477
14     22             d   bottom  250  58.670419
15     23             d      top  250  76.101363
16     24             a   bottom  275   74.41473
17     25             a      top  275  68.576858
18     26             b   bottom  275   61.19698
19     27             b      top  275  98.594881
20     28             c   bottom  275  92.893227
21     29             c      top  275  92.098049
22     30             d   bottom  275  51.747686
23     31             d      top  275   59.57162
24     32             a   bottom  300  98.786784
25     33             a      top  300   50.22847
26     34             b   bottom  300  99.248541
27     35             b      top  300  58.169349
28     36             c   bottom  300  65.473084
29     37             c      top  300  53.960273
30     38             d   bottom  300   56.18311
31     39             d      top  300  50.971626

### c. Which recipe, on average, is the best? recipe b

In [20]:
cake_melt.groupby('recipe_letter').score.mean()

recipe_letter
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: score, dtype: float64

### d. Which oven temperature, on average, produces the best results? 275

In [21]:
cake_melt.groupby('temp').score.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: score, dtype: float64

### e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees

In [29]:
cake_melt.groupby(['recipe_letter', 'position', 'temp']).score.mean().sort_values(ascending=False)

recipe_letter  position  temp
b              bottom    300     99.248541
a              bottom    300     98.786784
b              top       275     98.594881
d              top       225     96.873178
c              bottom    225     96.470207
b              top       250     95.224151
c              bottom    275     92.893227
               top       275     92.098049
                         250     82.795477
b              top       225     82.455004
d              top       250     76.101363
a              bottom    275     74.414730
c              top       225     71.306308
a              top       275     68.576858
c              bottom    300     65.473084
b              bottom    250     61.904369
a              bottom    225     61.738655
b              bottom    275     61.196980
d              top       275     59.571620
               bottom    250     58.670419
b              top       300     58.169349
               bottom    225     57.095320
d              bottom   